### DISCUSSION TITLE
connect using wallet_dir needs TNS_ADMIN set ?

### DISCUSSION THREAD

**Discussion by JohnvdL** [oracle/python-oracledb#74] [Q&A]
Hi,

I'm trying to connect to a DB using a wallet. To do this I use the 'thick mode', and provide the wallet_location when calling oracledb.connect:
```
import os
import oracledb as db_driver

ORACLE_HOME = "/Users/me/oracle_client"
os.environ["ORACLE_HOME"] = ORACLE_HOME

wallet_dir="/Users/me/dbs/DEVPDB_12345"
my_dsn="...this is a long string..."

db_driver.init_oracle_client()      # Thick mode
db_conn = db_driver.connect(dsn=my_dsn, wallet_location=wallet_dir)
```
This fails with 
```
oracledb.exceptions.DatabaseError: ORA-01017: invalid username/password; logon denied
```

However, if I set the environment variable TNS_ADMIN to the same value as the wallet_dir variable, before launching the  script the connection is created ok. 

What I'm trying to do is connecting to several DBs from within 1 script, with each DB having its own wallet. So setting TNS_ADMIN in the environment is not a solution. 
What am I doing wrong in opening a connection, using a wallet_location? 

Many thanks, regards,

JohnvdL

**Comment by cjbj** [comment]
For multiple wallets in general see https://github.com/oracle/python-cx_Oracle/issues/485#issuecomment-729843502

In thick mode, I advise using 21c for this; I've never managed to convince the powers-that-be to backport some fixes for multiple wallet support.

I will look at the specific of setting behavior later.

**Reply by anthony-tuininga** [reply]
Thin mode is capable of connecting with multiple wallets -- but you will need to either specify the `wallet_password` parameter or ensure the `ewallet.pem` file is stored unencrypted. Note that the `wallet_location` parameter is not usable in thick mode. Instead, you have to inject it into the connect string (as shown in the comment @cjbj mentioned). One additional option, though, is to supply all of the individual parameters ( but do NOT supply the `dsn` parameter) and then python-oracledb will build the connect string with the wallet location included automatically.

**Reply by JohnvdL** [reply]
Hi, thanks for the reply, and I'll definitely try the suggested solution and report back. For now I guess there's a 'bug' in the documentation, where it states for the connect method:

> The wallet_location parameter is expected to be a string that identifies the directory where the wallet can be found. In python-oracledb Thin mode, this must be the directory of the PEM-encoded wallet file, ewallet.pem. In python-oracledb Thick mode, this must be the directory of the file, cwallet.sso. This value is used in both the python-oracledb Thin and Thick modes.

Thanks & cheers

**Reply by anthony-tuininga** [reply]
The documentation is accurate, but perhaps not complete -- as the only way that `wallet_location` will be used in thick mode is if a dsn is *NOT* supplied (but only individual parameters are specified). When only individual parameters are supplied the DSN is built and then `wallet_location` (and other parameters) are usable in thick mode.

**Reply by cjbj** [reply]
An update: If I read the internal Oracle tracking tooling correctly, backports for the Oracle client library issues I mentioned earlier seem to have landed in the recently released 19.17 client libraries.
